iwyu.cc

```diff
@@ -1380,30 +1370,53 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
-  // To help us guess the owner, we use this rule: if the macro-file
-  // intends-to-provide the type, then we keep ownership of the type
-  // at the macro.  Otherwise, we assume it's with the caller.  This
-  // works well as long as the file defining the macro is well-behaved.
-  //   This function should be called with a use-loc that is within
-  // an expanded macro (so the use-loc will point to either inside a
-  // macro definition, or to an argument to a macro call).  If it
-  // points within a macro definition, and that macro-definition file
-  // does not mean to re-export the symbol being used, then we reassign
-  // use of the decl to the macro-caller.
+  // Return any forward-declare of decl earlier in the same file as loc.
+  const NamedDecl* GetForwardDeclareInSameFile(const Decl* decl,
+                                               SourceLocation loc) {
+    set<const NamedDecl*> redecls = GetClassRedecls(DynCastFrom(decl));
+    for (Each<const NamedDecl*> it(&redecls); !it.AtEnd(); ++it) {
+      if (IsBeforeInSameFile(*it, loc)) {
+        return *it;
+      }
+    }
+
+    return NULL;
+  }
+
+  // Assuming use_loc is part of a macro expansion, decide whether the file
+  // expanding the macro or the file defining the macro should be held
+  // responsible for the used_decl.
+  // The rules are:
+  // 1) If the macro definition file forward-declares the used decl, that's a
+  //    hint that it wants callers to take responsibility.
+  // 2) If the symbol being used is passed as an argument to the macro, we want
+  //    callers to take responsibility (that thing they passed in is something
+  //    they know about, and the macro doesn't)
+  // 3) If the use_loc is in <scratch space>, we assume it's formed by macro
+  //    argument concatenation, and attribute responsibility to callers.
+  // Otherwise, the macro file is responsible for including the symbol.
+  //
+  // Returns the responsible location.
   SourceLocation GetUseLocationForMacroExpansion(SourceLocation use_loc,
                                                  const Decl* used_decl) {
     CHECK_(IsInMacro(use_loc) && "Unexpected non-macro-expansion call");
-    if (!preprocessor_info().PublicHeaderIntendsToProvide(
-            GetFileEntry(use_loc), GetFileEntry(used_decl)))
-      return GetInstantiationLoc(use_loc);
+
+    SourceLocation caller_loc = GetInstantiationLoc(use_loc);
+    const char* responsible = "caller";
+
+    if (IsInScratchSpace(use_loc)) {
+      VERRS(5) << "Decl was used in <scratch space>, presumably as a result of "
+                  "macro arg concatenation.\n";
+      use_loc = caller_loc;
+    } else if (GlobalSourceManager()->isMacroArgExpansion(use_loc)) {
+      VERRS(5) << "Use location is a macro arg expansion.\n";
+      use_loc = caller_loc;
+    } else {
+      responsible = "macro";
+    }
+
+    VERRS(5) << "Attributing use to " << responsible << ".\n";
```

Comment:
Do comments have enough context? Especially in the case of attributing to macro it looks like the comment will be "Attributing use to macro". And it is unclear what use, to what macro.